pkg/pss/pss_test.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pss_test
+
+import (
+	"context"
+	"io/ioutil"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/pss"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	pushsyncmock "github.com/ethersphere/bee/pkg/pushsync/mock"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+// TestTrojanChunkRetrieval creates a trojan chunk
+// mocks the localstore
+// calls pss.Send method and verifies it's properly stored
+func TestTrojanChunkRetrieval(t *testing.T) {
+	var err error
+	ctx := context.TODO()
+	testTags := tags.NewTags()
+
+	// create a mock pushsync service to push the chunk to its destination
+	var receipt *pushsync.Receipt
+	var storedChunk swarm.Chunk
+	pushSyncService := pushsyncmock.New(func(ctx context.Context, chunk swarm.Chunk) (*pushsync.Receipt, error) {
+		rcpt := &pushsync.Receipt{
+			Address: swarm.NewAddress(chunk.Address().Bytes()),
+		}
+		storedChunk = chunk
+		receipt = rcpt
+		return rcpt, nil
+	})
+
+	pss := pss.New(logging.New(ioutil.Discard, 0), pushSyncService, testTags)
+
+	target := trojan.Target([]byte{1}) // arbitrary test target
+	targets := trojan.Targets([]trojan.Target{target})
+	payload := []byte("RECOVERY CHUNK")
+	topic := trojan.NewTopic("RECOVERY TOPIC")
+
+	// call Send to store trojan chunk in localstore
+	if _, err = pss.Send(ctx, targets, topic, payload); err != nil {
+		t.Fatal(err)
+	}
+
+	// create a stored chunk artificially
+	m, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var tc swarm.Chunk
+	tc, err = m.Wrap(targets)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tag, err := tags.NewTags().Create("pss-chunks-tag", 1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	storedChunk = tc.WithTagID(tag.Uid)
+
+	// check if receipt is received
+	if receipt == nil {
+		t.Fatal("receipt not received")
+	}
+
+	if !reflect.DeepEqual(tc, storedChunk) {
+		t.Fatalf("trojan chunk created does not match sent chunk. got %s, want %s", storedChunk.Address().ByteString(), tc.Address().ByteString())
+	}
+}
+
+// TestPssTags creates a trojan chunk
+// mocks the localstore
+// calls pss.Send method
+// updates the tag state (Stored/Sent/Synced)
```

Comment:
stored will not be incremented if  we  use the pushsync protocol directly

Reply:
fixed